### ISSUE TITLE
Deep copy operator

### ISSUE THREAD

**Issue by rhysd** [rhysd/Dachs#61]
I've not intended to implement user-defined deep copy operator.  However, this operator got mandatory because of pointer.  Copy of pointer doesn't copy its contents; shallow copy.  It means that user must define how to copy the contents of pointer.  Return type of the operator must be the same as its class.
Dachs's built-in types (e.g. `int`, `tuple`, `static_array`, `char`, ..., and of course `pointer`) are deep copied.

This operator is yielded throughout the program; where deep copy is required.

```
func main
    x := [1, 2, 3]

    # yield deep copy
    var y := x

    # yield deep copy
    y = x

    # yield deep copy
    foo(x)
end
```

I'm considering the name of operator.  I have three idea.  First, I may introduce operator `:=` like below:

```
class X
    ptr : pointer
    size : uint

    init(@size)
        @ptr := new pointer(char){@size}
    end

    func :=
        ret new X{@size}
    end
end
```

Second, I may introduce `deep_copy()` special member function.  This is Ruby's way (e.g. `to_s`) like below:

```
class X
    ptr : pointer
    size : uint

    init(@size)
        @ptr := new pointer(char){@size}
    end

    func deep_copy
        ret new X{@size}
    end
end
```

Third, adding special block may be good like `init()`.

```
class X
    ptr : pointer
    size : uint

    init(@size)
        @ptr := new pointer(char){@size}
    end

    deep_copy
        ret new X{@size}
    end
end
```

Pros and Cons are:

- The first idea
  - __Pros__ : It doesn't pollute member function's name space.
  - __Cons__ : It is odd because not only `:=`(initialization) yields deep copy.  It is yielded in substitution, pass by value and (maybe) returned value.

- The second idea
  - __Pros__ : The name is reasonable and reusable.  User can also yields deep copy the object explicitly without any extra member function definition.
  - __Cons__ : `deep_copy` is very generic name.  User may want to use it for other purpose. What matters worse, it may be unintentionally called when user doesn't know the special function.  Changing to a name like `__deep_copy` which is harder to conflict (Python's way) may be helpful, but it doesn't solve the problem truly.

- The third idea
  - __Pros__ : This resolves above two conses.
  - __Cons__ : It may be overkill for the operator which is rarely used by general users.  The feature is for librarian.

**Comment by rhysd** [comment]
I implemented with second idea in [wip/issue/61](https://github.com/rhysd/Dachs/tree/wip/issue/61) branch.
All AST nodes which may invoke deep copy has a member variable of `scope::weak_func_scope` to remember which function should be invoked for deep copy.

However, it has a problem when the data structure is nested. For example, when the value of `[X(int)]` is copied, it copies the whole array with default deep copy and never invoke user defined deep copy operator for its element `X(int)`.
I must fix it but it requires to change the way to have a function of deep copy.

**Comment by rhysd** [comment]
I re-implemented with solving above problem.  I employed 3rd idea (change keyword `deep_copy` to `copy`).

```
class X
    ptr : pointer
    size : uint

    init(@size)
        @ptr := new pointer(char){@size}
    end

    copy
        ret new X{@size}
    end
end
```

I introduced `copy ... end` special block because only one copy function should exist per one type.  If previous idea is continued, the copy behavior may change in different name spaces.  It's evil.  Different copy function may be invoked in different place; it means that all copy in code must know what copy function is invoked at the place.  If `copy ... end` is introduced, the same copy function will be invoked wherever the invoked point is for the same type because only one function can be defined for one type.

In addition, I intend to add `move ... end` to move an object to improve performance even if it is in value semantics.
However, move is less often defined than copy because move is often the shallow copy of an object and default copy does it.
